src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -0,0 +1,240 @@
+    {
+        if (_queuedConnectionsCounter.Enabled)
+        {
+            ConnectionQueuedStopCore(connection);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        void ConnectionQueuedStopCore(BaseConnectionContext connection)
+        {
+            _queuedConnectionsCounter.Add(-1, CreateTagList(connection));
+        }
+    }
+
+    public void RequestQueuedStart(BaseConnectionContext connection, string httpVersion)
+    {
+        if (_queuedRequestsCounter.Enabled)
+        {
+            RequestQueuedStartCore(connection, httpVersion);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        void RequestQueuedStartCore(BaseConnectionContext connection, string httpVersion)
+        {
+            var tags = CreateTagList(connection);
+            tags.Add("version", httpVersion);
+            _queuedRequestsCounter.Add(1, tags);
+        }
+    }
+
+    public void RequestQueuedStop(BaseConnectionContext connection, string httpVersion)
+    {
+        if (_queuedRequestsCounter.Enabled)
+        {
+            RequestQueuedStopCore(connection, httpVersion);
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        void RequestQueuedStopCore(BaseConnectionContext connection, string httpVersion)
+        {
+            var tags = CreateTagList(connection);
+            tags.Add("version", httpVersion);
+            _queuedRequestsCounter.Add(-1, tags);
+        }
+    }
+
+    public void RequestUpgradedStart(BaseConnectionContext connection)
+    {
+        if (_currentUpgradedRequestsCounter.Enabled)
+        {
+            RequestUpgradedStartCore(connection);
+        }
+
+        void RequestUpgradedStartCore(BaseConnectionContext connection)
+        {
+            _currentUpgradedRequestsCounter.Add(1, CreateTagList(connection));
+        }
+    }
+
+    public void RequestUpgradedStop(BaseConnectionContext connection)
+    {
+        if (_currentUpgradedRequestsCounter.Enabled)
+        {
+            RequestUpgradedStopCore(connection);
+        }
+
+        void RequestUpgradedStopCore(BaseConnectionContext connection)
+        {
+            _currentUpgradedRequestsCounter.Add(-1, CreateTagList(connection));
+        }
+    }
+
+    private static TagList CreateTagList(BaseConnectionContext connectionContext)
+    {
+        var tags = new TagList();
+        if (connectionContext.LocalEndPoint is { } localEndpoint)
+        {
+            tags.Add("endpoint", localEndpoint.ToString());
+        }
+
+        return tags;
+    }
+
+    public void TlsHandshakeStart()
+    {
+        _currentTlsHandshakesCounter.Add(1);
+    }
+
+    public void TlsHandshakeStop(long startTimestamp, long currentTimestamp, string? exceptionName = null)
+    {
+        _currentTlsHandshakesCounter.Add(-1);
+
+        var tags = new TagList();
+        if (exceptionName != null)
+        {
+            tags.Add("exception-name", exceptionName);
```

Comment:
Do we want some standardization across the .NET repo? (assuming we expect other places to have the same "exception-name" metrics dimension) @noahfalk FYI

Reply:
I'm curious if the user would end up seeing many AggregatedExceptions (which might be less useful)?

Reply:
Yes, that's possible. Maybe AggregatedExceptions could be transformed into a comma-separated list of exception names?

I want something more useful than a bool `has-error`, but is still low cardinality. Exception name feels like the best choice.

Reply:
> Maybe AggregatedExceptions could be transformed into a comma-separated list of exception names? Or maybe even multiple `exception-name` tags.

This might cause a cardinality explosion.

Reply:
Stepping back question regarding the direction - do we want to have the exception name as a dimension at all? Do we think the direction should be Exemplar? Do we guarantee that exception names are stable (e.g. a future version of the ASP.NET Core won't change by throwing a different type of exception so folks can rely on this dimension for alerting)? If we don't have good answers for these questions - do we still want to include this dimension or maybe wait till we have better understanding?

Reply:
yes